[#8] 이메일을 활용한 비밀번호 찾기 구현

msugo1 opened the pull request:
## 📖 구현 항목
 ###  🔖 유저 아이디 및 비밀번호 확인
- 이메일 전송에 앞서 전달받은 요청 정보를 활용해 DB에서 `사용자 존재유무`와 `이메일 일치` 확인
- 사용자가 존재하지 않거나 이메일이 일치하지 않는 경우에는 예외 발생

 ###  🔖 임시비밀번호 및 메일 내용을 담을 객체를 만들 클래스 작성
- 위의 검증과정을 통과하면 `UUID`를 이용해 사용자에게 전달해줄 임시 비밀번호를 생성 
- 발급한 임시비밀번호를 포함해 메일의 제목과 내용을 담을 수 있을 객체를 생성하기 위한 클래스 생성

 ###  🔖 메일 발송기능
- 위에서 담은 내용을 이메일로 만들어 발송해 줄 `EmailService` 클래스 작성

 ###  🔖 위의 기능을 테스트하기 위한 테스트 작성

Commit:
[#8] 사용자 비밀번호 찾기 구현 (1차)

Model 관련

* UserFindPasswordRequest (클래스 추가)
- 비밀번호를 찾기 위한 사용자의 요청을 담을 클래스 생성(userId와 email을
  필요로함)

Controller 관련

* UserController (메소드 추가)
- userFindPassword
: 사용자의 요청을 서비스 계층으로 전달해줄 메소드 추가

Service 관련

* UserService (메소드 추가)
- userFindPassword
(logic: 요청을 받으면 해당 email로 새롭게 초기화된 비밀번호를 발송해서
사용자가 로그인할 수 있도록 함)

1. 사용자의 요청으로 들어온 객체를 활용해 id와 email의 존재 여부를
   파악한 뒤 존재하지 않는 경우 예외발생
2. id, email이 모두 존재하는 경우 랜덤한 password를 생성한 뒤 email과 함께 EmailService로 전달해서 이메일을 발송할 수 있도록 함
3. 변경된 비밀번호는 DB에 반영해서 사용자가 새롭게 발송된 비밀번호를
   통해 로그인 할 수 있도록 함

* EmailService (클래스 및 메소드 새롭게 추가)
- sendEmailWithNewPassword
: UserService에서 전달 받은 이메일과 패스워드를 이용해 사용자에게 해당
이메일로 임시 비밀번호를 발송

Exception 관련

* InvalidEmailException (추가)
- 이메일이 존재하지 않는 경우 맞는 에외를 발생시키기 위한 예외 클래스

* ControllerExceptionHandler
- 위의 예외를 핸들러에 추가하여 Controller 계층에서 try-catch로
  예외처리를 하지 않아도 자동으로 이루어지도록 함

Commit:
Merge branch 'feature/6' into feature/8

Commit:
[#8] 이메일을 이용한 사용자 비밀번호 찾기 기능 구현

(update)

- 임시 비밀번호를 포함한 `email`에 담을 내용을 `UserService`에서 생성하여 `EmailService`으로 전달하여 EmailService는 이메일 내용의 구성과 전송만을 담당하도록 변경

- EmailService 내부에 메일 내용을 고정해 놓는 대신 외부에서 내용을
  주입받는 형식으로 변경해서 코드의 중복 없이 좀 더 유연하게 메일 내용을
  구성하고 보낼 수 있도록 함

- 예외 복구가 불가능해 보이는 checked 예외인 `MessagingException`과 이미 Runtime 예외인 `MailException`을 예외처리하여 예외 발생 시 Controller 계층에서도 이를 handling 할 수 있도록함 (`MailNotSentException` - 예외 클래스 새로 생성)

- 이메일의 내용을 구성하는 기본 틀을 담은 클래스
- 본문은 차후에 각 목적에 맞는 클래스를 생성하여 EmailBase를 상속받고
  `writeBody` 메소드를 오버라이딩함으로써 비밀번호 찾기가 아닌 다른
  목적의 메일 작성에도 활용할 가능성을 만듦.

- EmailBase를 상속받은 자식 클래스
- 제목과 본문을 담고 있으며 임시비밀번호를 담아서 전송하는 틀을 가지고
  있음

- 위에서 구현한 UserService의 비밀번호 찾기 로직을 테스트하기 위해
  테스트 작성

Commit:
[#8] 오탈자 수정

- 주석에 있는 오탈자를 수정

f-lab-dev commented on src/main/java/me/soo/helloworld/controller/UserController.java:
```diff
@@ -54,4 +54,12 @@
+
+        userService.findUserPassword(findPasswordRequest);
+
+        return HTTP_RESPONSE_OK;
```
리턴타입을 void로 하면 이것은 생략이 가능합니다~

msugo1 replied:
리턴타입 `ResponseEntity<Void> -> void`로 바꾸고 리턴값 생략하였습니다.

f-lab-dev commented on src/main/java/me/soo/helloworld/model/email/EmailFindPassword.java:
```diff
@@ -0,0 +1,17 @@
+        this.body = writeBody(newPassword);
+    }
+
+    String writeBody(String newPassword) {
```
접근제한자를 설정하지 않으신 이유가 있으실까요?

msugo1 replied:
접근제한자를 설정하지 않았던 것은 기존 구상과 관련이 있습니다.

기존에 구상했던 내용은 `EmailBase`를 추상 클래스로 두고 이를 상속받은 클래스에서 목적에 맞게 `writeBody`를 오버라이딩 하여 본문 내용을 암묵적으로 만들어내는 것이었습니다.

따라서, `writeBody` 메소드는 단순히 해당 클래스 내에서만 쓰일 용도로 작성되었기 때문에 접근제한자가 `private`이 되어야 맞습니다. 

그러나 해당 메소드는 부모 클래스에서  `abstract`로 선언된 메소드이며, 상속 받은 클래스만 내부에서 사용할 수 있도록 하기 위해서 접근제한자를 따로 두지 않고 메소드를 작성했습니다.

자식 클래스들이 부모 클래스에 있는 메소드를 오버라이딩 한 경우 접근제한자가 제한하는 범위가 더 좁아질 수 없습니다.  (같거나 더 넓어져야 합니다.) 때문에 `private`으로 설정할 수 없었고, 가능한 최소 범위로 만들기 위해 접근제한자를 설정하지 않았습니다.

f-lab-dev commented on src/main/java/me/soo/helloworld/model/email/EmailFindPassword.java:
```diff
@@ -0,0 +1,17 @@
+
+    String writeBody(String newPassword) {
+        StringBuilder body = new StringBuilder();
+        body.append("회원님의 임시 비밀번호는 ")
```
이런곳에는 `String.format`을 쓰는것도 괜찮습니다~

msugo1 replied:
자바 성능 튜닝 이야기볼 때 `formatter`가 성능에 문제를 끼친다고 해서 안썼었는데 다시보니까 `System.out.foramt()`을 의미하는 거였네요.
`String.format` 알아보겠습니다. 감사합니다.

f-lab-dev commented on src/main/java/me/soo/helloworld/model/email/EmailFindPassword.java:
```diff
@@ -0,0 +1,17 @@
+        body.append("회원님의 임시 비밀번호는 ")
+                .append(newPassword)
+                .append("입니다. 로그인 후 비밀번호를 다시 변경해 주세요.");
+        return String.valueOf(body);
```
이렇게 하기보단 StringBuilder의 메소드를 활용하시면 좋을 것 같네요

f-lab-dev commented on src/main/java/me/soo/helloworld/model/email/EmailFindPassword.java:
```diff
@@ -0,0 +1,17 @@
+
+    public EmailFindPassword(String newPassword) {
+        this.title = "<Hello World> 회원님의 임시비밀번호 안내입니다.";
+        this.body = writeBody(newPassword);
```
객체를 생성하기만해도 내부 설정이 다 완료되니 편해보이기는 하는데 동작이 너무 암묵적입니다.

EmailBase는 값만 담을 수 있도록 해두고, 이런 메일을 발송하거나 내용을 build해주는 클래스를 나눠보면 좋을 것 같아요

f-lab-dev commented on src/main/java/me/soo/helloworld/service/EmailService.java:
```diff
@@ -0,0 +1,45 @@
+            mailSender.send(message);
+
+        } catch (MessagingException | MailException e) {
+            throw new MailNotSentException("메일 시스템에 문제가 있어 해당 이메일 발송에 실패하였습니다. 잠시 후에 다시 시도해 주세요.");
```
원활한 디버깅을 위해 `cause`에 `e`를 넣어서 던져도 좋을 것 같습니다.

msugo1 replied:
피드백 반영했습니다!

f-lab-dev commented:
반영된 커밋로그가 안보이는데 혹시 push하신걸까요?

Commit:
[#8] 피드백 반영

- UserController 내 userFindPassword의 메소드 리턴 타입을
  `ResponseEntity<Void> -> void`로 변경하여 리턴 값 생략

- 기존: EmailBase를 상속받은 EmailFindPassword 클래스의
  객체로 새 비밀번호를 파라미터로 넘겨주면 메일 내용을 담은 객체가 자동
  생성되도록 구성

- 편리하지만 동작이 너무 암묵적

- 변경
1. EmailBuilder 클래스 생성 및 build 메소드 추가
: build 메소드는 보낼 사람(to), 제목(title), 내용(body)를 파라미터로
받고 해당 값을 담은 EmailBase 객체를 리턴하도록 변경

- MailService 클래스 내 sendEmail 메소드에서 catch로 예외 처리할 때 원할한 디버깅을
  위해 예외 메시지와 함께 cause 객체인 'e'도 함께 전달하도록 변경

f-lab-dev commented on src/main/java/me/soo/helloworld/service/UserService.java:
```diff
@@ -79,10 +79,14 @@ public void findUserPassword(UserFindPasswordRequest findPasswordRequest) {
         String newPassword = UUID.randomUUID().toString();
-        EmailBase emailContent = new EmailFindPassword(newPassword);
-        emailService.sendEmail(findPasswordRequest.getEmail(), emailContent);
+        String title = "임시 비밀번호 안내입니다.";
```
이런 값들은 상수로 처리해줘도 될 것 같습니다~

f-lab-dev commented on src/test/java/me/soo/helloworld/service/UserServiceTest.java:
```diff
@@ -137,8 +137,12 @@ public void findUserPasswordSuccess() {
-        doNothing().when(emailService).sendEmail(userFromDB.getEmail(), emailContent);
+
+        String title = "임시 비밀번호 안내입니다.";
+        String content = String.format("회원님의 임시 비밀번호는 %s 입니다. 로그인 후 비밀번호를 변경해주세요", newPassword);
```
여기 또한 위의 내용과 중복인데 상수로 처리하면 공통화시킬 수 있을 것 같네요~

Commit:
[#8] 피드백 반영

* userFindPassword
- 하드코딩한 부분(이메일 제목 및 내용)을 상수로 변경

f-lab-dev commented on src/main/java/me/soo/helloworld/util/EmailBuilder.java:
```diff
@@ -4,12 +4,18 @@
+
+    public final static String FIND_PASSWORD_BODY = "회원님의 임시 비밀번호 입니다. 로그인 후 비밀번호를 변경해주세요";
+
+    public static EmailBase build(String to, String title, String content, String body) {
```
빌더패턴을 적용하시려고 한 것 같은데 빌더패턴에 대해 다시 한번 찾아보시면 좋을 것 같아요~

msugo1 replied:
빌더패턴 적용 할 때 코드에 나오는 값들

```
EmailBuilder.builder( )
    .to(emailAddress)
    .title(emailTitle)
    .body(emailBody)
.build( );
```

이 비즈니스 로직에 있으니 조금 지저분해보여서 단순화하려고 했다가 결국 비슷하게 만들어버렸네요.

빌더패턴 다시 찾아보고 다시 몇 시간 생각하고 고민 끝에 결국 조금 수정하긴 했는데 잘 되었는지 모르겠네요.

Commit:
[#8] 피드백 반영

- EmailBase 객체 생성에 있어서 `Build Pattern`을 다른 방향으로 적용

f-lab-dev commented on src/main/java/me/soo/helloworld/model/email/FindPasswordEmail.java:
```diff
@@ -0,0 +1,17 @@
+    static public EmailBase create(String to, String temporaryPassword) {
+
+        return EmailBase.builder()
+        .to(to)
```
이런곳엔 들여쓰기를 넣어주시는게 가독성 향상에 좋습니다~

Commit:
[#8] 피드백 반영

* FindPasswordEmail
- 가독성 향상을 위한 들여쓰기

msugo1 committed:
Merge branch 'develop' into feature/8
